server/src/main/java/org/elasticsearch/common/bytes/PagedBytesReference.java

```diff
@@ -39,11 +39,18 @@
     private final int offset;
     private final int length;
 
-    public PagedBytesReference(ByteArray byteArray, int length) {
-        this(byteArray, 0, length);
+    public static BytesReference of(ByteArray byteArray, int length) {
+        if (length == 0) {
+            return BytesArray.EMPTY;
+        }
+        if (byteArray.hasArray()) {
+            return new BytesArray(byteArray.array(), 0, length);
```

Comment:
Why have another step going to the `byte[]` through `ByteArray` when we don't have to, especially with the new optimizations for that case in here.